README.md
# Dyno Metadata

[![GitHub Build Status](https://github.com/dentarg/dyno_metadata/workflows/Ruby/badge.svg)](https://github.com/dentarg/dyno_metadata/actions)

Helpers to access [Heroku Dyno Metadata](https://devcenter.heroku.com/articles/dyno-metadata) from [`ENV`](https://ruby-doc.org/core-2.2.0/ENV.html). Graceful fallback to dummy values (useful in development).

Installation

    gem install dyno_metadata

Methods

```ruby
$ ruby -I lib -r dyno_metadata.rb -e ' \
(DynoMetadata.methods - Object.methods).each { |method| \
  puts "DynoMetadata.#{method}\n" ; \
}'
DynoMetadata.app_id
DynoMetadata.app_name
DynoMetadata.dyno_id
DynoMetadata.release_created_at
DynoMetadata.release_version
DynoMetadata.slug_commit
DynoMetadata.commit
DynoMetadata.slug_description
DynoMetadata.short_commit
DynoMetadata.to_h
```

Demonstration

```ruby
$ ruby -I lib -r dyno_metadata.rb -e ' \
(DynoMetadata.methods - Object.methods).each { |method| \
  puts "> DynoMetadata.#{method}\n" ; \
  print "=> " ; \
  p DynoMetadata.public_send(method) \
} ; \
puts "> DynoMetadata.short_commit(12)\n" ; \
print "=> " ; \
p DynoMetadata.short_commit(12)'
> DynoMetadata.app_id
=> "9daa2797-e49b-4624-932f-ec3f9688e3da"
> DynoMetadata.app_name
=> "example-app"
> DynoMetadata.dyno
=> "web.1"
> DynoMetadata.dyno_id
=> "1vac4117-c29f-4312-521e-ba4d8638c1ac"
> DynoMetadata.release_created_at
=> "2015-04-02T18:00:42Z"
> DynoMetadata.release_version
=> "v42"
> DynoMetadata.slug_commit
=> "2c3a0b24069af49b3de35b8e8c26765c1dba9ff0"
> DynoMetadata.commit
=> "2c3a0b24069af49b3de35b8e8c26765c1dba9ff0"
> DynoMetadata.slug_description
=> "Deploy 2c3a0b2"
> DynoMetadata.short_commit
=> "2c3a0b2"
> DynoMetadata.to_h
=> {:app_id=>"9daa2797-e49b-4624-932f-ec3f9688e3da", :app_name=>"example-app", :dyno=>"web.1", :dyno_id=>"1vac4117-c29f-4312-521e-ba4d8638c1ac", :release_created_at=>"2015-04-02T18:00:42Z", :release_version=>"v42", :slug_commit=>"2c3a0b24069af49b3de35b8e8c26765c1dba9ff0", :slug_description=>"Deploy 2c3a0b2", :short_commit=>"2c3a0b2"}
> DynoMetadata.short_commit(12)
=> "2c3a0b24069a"
```

## Development

### Release workflow

* Update the demonstration in this README if needed.

* Bump the version in `version.rb` in a commit, no need to push (the release task does that).

* Build and [publish](http://guides.rubygems.org/publishing/) the gem. This will create the proper tag in git, push the commit and tag and upload to RubyGems. Use [keycutter](https://github.com/joshfrench/keycutter) to manage multiple RubyGems accounts.

        bundle exec rake release

    * If you are not logged in as on the correct account, the rake task will fail and tell you to set credentials via `gem push`, do that and run the `release` task again. Use [keycutter](https://github.com/joshfrench/keycutter) to manage multiple RubyGems accounts.

* Update the changelog manually, commit and push.


CHANGELOG.md
# Change Log

## Next

## [v0.0.4](https://github.com/dentarg/dyno_metadata/tree/v0.0.4) (2022-11-03)
[Full Changelog](https://github.com/dentarg/dyno_metadata/compare/v0.0.3...v0.0.4)

* Support alternative `ENV`s [`475d005`](https://github.com/dentarg/dyno_metadata/commit/475d005)
* Bump required Ruby version to 2.5.9 [`e0efd36`](https://github.com/dentarg/dyno_metadata/commit/e0efd36)

## [v0.0.3](https://github.com/dentarg/dyno_metadata/tree/v0.0.2) (2021-11-20)
[Full Changelog](https://github.com/dentarg/dyno_metadata/compare/v0.0.2...v0.0.3)

* Require MFA for RubyGems operations [`1b41bbe`](https://github.com/dentarg/dyno_metadata/commit/1b41bbe)
* Use GitHub Actions for CI [#7](https://github.com/dentarg/dyno_metadata/pull/7)
* Add method to get `$DYNO` [`679b3a6`](https://github.com/dentarg/dyno_metadata/commit/679b3a6)
* Allow dynamic length for `short_commit` [`fb1c131`](https://github.com/dentarg/dyno_metadata/commit/fb1c131)

## [v0.0.2](https://github.com/dentarg/dyno_metadata/tree/v0.0.2) (2016-12-08)
[Full Changelog](https://github.com/dentarg/dyno_metadata/compare/v0.0.1...v0.0.2)

**Closed issues:**

- Do not support Ruby 1.9.3 [\#1](https://github.com/dentarg/dyno_metadata/issues/1)

## [v0.0.1](https://github.com/dentarg/dyno_metadata/tree/v0.0.1) (2016-12-08)


spec/spec_helper.rb
# frozen_string_literal: true

require "climate_control"
require "securerandom"

RSpec.configure do |conf|
  conf.disable_monkey_patching!
  conf.filter_run :focus
  conf.order = :random
  conf.run_all_when_everything_filtered = true
  conf.warnings = true
end


spec/unit/dyno_metadata_spec.rb
# frozen_string_literal: true

RSpec.describe DynoMetadata do
  let(:defaults) do
    {
      app_id:             "9daa2797-e49b-4624-932f-ec3f9688e3da",
      app_name:           "example-app",
      dyno:               "web.1",
      dyno_id:            "1vac4117-c29f-4312-521e-ba4d8638c1ac",
      release_created_at: "2015-04-02T18:00:42Z",
      release_version:    "v42",
      slug_commit:        "2c3a0b24069af49b3de35b8e8c26765c1dba9ff0",
      slug_description:   "Deploy 2c3a0b2",
      short_commit:       "2c3a0b2",
    }
  end

  subject { described_class.to_h }
  it { is_expected.to match(defaults) }

  context "values from ENV" do
    test_env = {
      HEROKU_APP_ID:             SecureRandom.hex,
      HEROKU_APP_NAME:           SecureRandom.hex,
      DYNO:                      SecureRandom.hex,
      HEROKU_DYNO_ID:            SecureRandom.hex,
      HEROKU_RELEASE_CREATED_AT: SecureRandom.hex,
      HEROKU_RELEASE_VERSION:    SecureRandom.hex,
      HEROKU_SLUG_COMMIT:        SecureRandom.hex,
      HEROKU_SLUG_DESCRIPTION:   SecureRandom.hex,
    }

    around do |example|
      ClimateControl.modify(test_env) do
        example.run
      end
    end

    subject { described_class.to_h.tap { |hsh| hsh.delete(:short_commit) }.values }
    it { is_expected.to eq(test_env.values) }
  end

  context "alternative values from ENV" do
    let(:test_env) do
      {
        # https://fly.io/docs/reference/runtime-environment/
        FLY_APP_NAME:       SecureRandom.hex,
        FLY_ALLOC_ID:       "b996131a-5bae-215b-d0f1-2d75d1a8812b",
        FLY_REGION:         "ams",
        # Not documented, found with "fly ssh console"
        FLY_PUBLIC_IP:      "2605:4c40:197:7fd0:0:fa4a:e57b:1",
        FLY_VCPU_COUNT:     "1",
        FLY_VM_MEMORY_MB:   "256",
        # https://github.com/dentarg/fly/commit/b8685c570ea849367f1a719121bcba0bf6645b3f
        RELEASE_CREATED_AT: "2022-11-05T17:29:22+00:00",
        RELEASE_VERSION:    "v123",
        RELEASE_COMMIT:     "e0efd36c5f678ace3c54694b7bf94db6ba4cb649",
      }
    end

    around do |example|
      ClimateControl.modify(test_env) do
        example.run
      end
    end

    describe ".app_name" do
      subject { described_class.app_name }
      it { is_expected.to eq(test_env.fetch(:FLY_APP_NAME)) }
    end

    describe ".dyno_id" do
      subject { described_class.dyno_id }
      it { is_expected.to eq(test_env.fetch(:FLY_ALLOC_ID)) }
    end

    describe ".release_created_at" do
      subject { described_class.release_created_at }
      it { is_expected.to eq(test_env.fetch(:RELEASE_CREATED_AT)) }
    end

    describe ".release_version" do
      subject { described_class.release_version }
      it { is_expected.to eq(test_env.fetch(:RELEASE_VERSION)) }
    end

    describe ".release_commit" do
      subject { described_class.release_commit }
      it { is_expected.to eq(test_env.fetch(:RELEASE_COMMIT)) }
    end

    describe ".commit" do
      subject { described_class.commit }
      it { is_expected.to eq(test_env.fetch(:RELEASE_COMMIT)) }
    end

    describe ".short_commit" do
      subject { described_class.short_commit }
      it { is_expected.to eq(test_env.fetch(:RELEASE_COMMIT)[0..6]) }
    end

    describe ".fly_alloc_id" do
      subject { described_class.fly_alloc_id }
      it { is_expected.to eq(test_env.fetch(:FLY_ALLOC_ID)) }
    end

    describe ".fly_region" do
      subject { described_class.fly_region }
      it { is_expected.to eq(test_env.fetch(:FLY_REGION)) }
    end

    describe ".fly_public_ip" do
      subject { described_class.fly_public_ip }
      it { is_expected.to eq(test_env.fetch(:FLY_PUBLIC_IP)) }
    end

    describe ".fly_vcpu_count" do
      subject { described_class.fly_vcpu_count }
      it { is_expected.to eq(test_env.fetch(:FLY_VCPU_COUNT)) }
    end

    describe ".fly_vm_memory_mb" do
      subject { described_class.fly_vm_memory_mb }
      it { is_expected.to eq(test_env.fetch(:FLY_VM_MEMORY_MB)) }
    end
  end

  describe ".commit" do
    subject { described_class.commit }
    it { is_expected.to eq(defaults.fetch(:slug_commit)) }
  end

  describe ".short_commit(12)" do
    subject { described_class.short_commit(12) }
    it { is_expected.to eq("2c3a0b24069a") }
  end

  describe "::VERSION" do
    subject { described_class::VERSION }
    it { is_expected.to match(/\d+.\d+.\d+/) }
  end

  describe ".fetch" do
    it "is private" do
      expect { described_class.fetch("FOO", "BAR") }.to raise_error(/private/)
    end
  end
end


lib/dyno_metadata/dyno_metadata.rb
# frozen_string_literal: true

module DynoMetadata
  module_function

  def app_id
    fetch "HEROKU_APP_ID", "9daa2797-e49b-4624-932f-ec3f9688e3da"
  end

  def app_name
    fetch "HEROKU_APP_NAME", "FLY_APP_NAME", "example-app"
  end

  def dyno
    fetch "DYNO", "web.1"
  end

  def dyno_id
    fetch "HEROKU_DYNO_ID", "FLY_ALLOC_ID", "1vac4117-c29f-4312-521e-ba4d8638c1ac"
  end

  def fly_alloc_id
    fetch "FLY_ALLOC_ID", "b996131a-5bae-215b-d0f1-2d75d1a8812b"
  end

  def fly_region
    fetch "FLY_REGION", "ams"
  end

  def fly_public_ip
    fetch "FLY_PUBLIC_IP", "127.0.0.1"
  end

  def fly_vcpu_count
    fetch "FLY_VCPU_COUNT", "99"
  end

  def fly_vm_memory_mb
    fetch "FLY_VM_MEMORY_MB", "1337"
  end

  def release_created_at
    fetch "HEROKU_RELEASE_CREATED_AT", "RELEASE_CREATED_AT", "2015-04-02T18:00:42Z"
  end

  def release_version
    fetch "HEROKU_RELEASE_VERSION", "RELEASE_VERSION", "v42"
  end

  def slug_commit
    fetch "HEROKU_SLUG_COMMIT", "RELEASE_COMMIT", "2c3a0b24069af49b3de35b8e8c26765c1dba9ff0"
  end
  singleton_class.send(:alias_method, :commit, :slug_commit)
  singleton_class.send(:alias_method, :release_commit, :slug_commit)

  def slug_description
    fetch "HEROKU_SLUG_DESCRIPTION", "Deploy 2c3a0b2"
  end

  def short_commit(length = 7)
    slug_commit[0, length]
  end

  def to_h # rubocop:disable Metrics/MethodLength
    {
      app_id:             app_id,
      app_name:           app_name,
      dyno:               dyno,
      dyno_id:            dyno_id,
      release_created_at: release_created_at,
      release_version:    release_version,
      slug_commit:        slug_commit,
      slug_description:   slug_description,
      short_commit:       short_commit,
    }
  end

  private_class_method def fetch(*variable_names, fallback)
    variable_names.each do |var_name|
      return ENV[var_name] if ENV.key?(var_name)
    end

    fallback
  end
end
